### PR TITLE
SRC-134: Expose pre-filters to the public API

### PR DESCRIPTION
**Pull request by JamesPace-GBI**
# DO NOT MERGE

Now Bridge supports the pre-filter string body parameter we should include this in the Java API Client.

### CODE REVIEW

**Comment by JamesPace-GBI** [comment]
Maybe once the public docs have been updated I could include a link to them, rather than just mentioning them?

File: src/main/java/com/groupbyinc/api/Query.java [hunk 1]
```diff
@@ -1009,6 +1011,46 @@ public Query setLanguage(String language) {
...
+   * <code>
+   * Sets the pre-filter expression on the query and restricts the results based on the logical conditions defined by
+   * the expression. Fields targeted by the pre-filter expression must be refinable in the upload configuration.
+   *
+   * For example, if your records have a refinable field called "brand" and you wish to pre-filter the search for the brands
+   * "Samsung" and "Sony" your pre-filter expression would look like:
+   *
+   * `brand IN ["Samsung", "Sony"]`
+   *
+   * If you wanted to enhance this pre-filter and your records have a refinable field called "price", with a nested field called
```
Comment (HaiHoang-GBI) [inline src/main/java/com/groupbyinc/api/Query.java:1031]: @JamesPace-GBI , side question:What happens if somebody tries to do a pre-filter on an non-refinable field?

File: src/test/java/com/groupbyinc/api/QueryTest.java [hunk 2]
```diff
@@ -576,4 +576,20 @@ public void testPinnedRefinementsWithoutValueRefinements() {
                 "    'skip': 0" + //
                 "}", test);
   }
+
+  @Test
+  public void testPreFilterExpression() {
+    test.setQuery("nice products");
+    test.setPreFilterExpression("brand = \"shiny\"");
```
Comment (HaiHoang-GBI) [inline src/test/java/com/groupbyinc/api/QueryTest.java:583]: Another side question, do we support '  or just
i.e. "brand = 'shiny'"
vs "brand = \"shiny\""

**Review by HaiHoang-GBI — DISMISSED** [review]
LGTM. See comments/questions

File: src/test/java/com/groupbyinc/api/QueryTest.java (hunk 2 above)
Comment (HaiHoang-GBI) [reply src/test/java/com/groupbyinc/api/QueryTest.java:583]: Answering my own question.
From the API-javascript repos, I can see why we may just support the ".
In javascript, most people are use to surrounding literals with '
.withPreFilterExpression('brand = "shiny"')

File: src/main/java/com/groupbyinc/api/Query.java (hunk 1 above)
Comment (JamesPace-GBI) [reply src/main/java/com/groupbyinc/api/Query.java:1031]: I think they'll end up getting a 500 Internal Server Error, as it manifests as some variant of an Elasticsearch error. They should hopefully only get these errors when they are building the expression for the first time, unless they change the upload configuration of course.

File: src/test/java/com/groupbyinc/api/QueryTest.java (hunk 2 above)
Comment (JamesPace-GBI) [reply src/test/java/com/groupbyinc/api/QueryTest.java:583]: Yeah that's right. In the JavaScript project there are linting rules that only allow you to use single quotes for strings.

**Review by archerz — DISMISSED** [review]
LGTM.

**Comment by archerz** [comment]
> Maybe once the public docs have been updated I could include a link to them, rather than just mentioning them?

Yes. https://groupbysupport.docs.stoplight.io/gb-docs/data-and-upload-management/search/pre-filters

**Comment by JamesPace-GBI** [comment]
> > Maybe once the public docs have been updated I could include a link to them, rather than just mentioning them?
> 
> Yes. https://groupbysupport.docs.stoplight.io/gb-docs/data-and-upload-management/search/pre-filters

Done 👍

**Review by archerz — APPROVED** [review]

**Pull request merged**